In [20]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_svmlight_file

# Madelon dataset

## Loading the dataset

### Training set

In [39]:
def get_data(file):
    data = load_svmlight_file(file)
    return data[0], data[1]

X, y = get_data("madelon.txt")
print(X)

  (0, 5)	1.0
  (0, 7)	1.0
  (0, 14)	1.0
  (0, 20)	1.0
  (0, 28)	1.0
  (0, 32)	1.0
  (0, 33)	1.0
  (0, 36)	1.0
  (0, 41)	1.0
  (0, 49)	1.0
  (0, 52)	1.0
  (0, 56)	1.0
  (0, 66)	1.0
  (0, 75)	1.0
  (0, 77)	1.0
  (0, 80)	1.0
  (0, 83)	1.0
  (0, 85)	1.0
  (0, 92)	1.0
  (0, 102)	1.0
  (0, 110)	1.0
  (1, 5)	1.0
  (1, 7)	1.0
  (1, 19)	1.0
  (1, 20)	1.0
  :	:
  (8122, 87)	1.0
  (8122, 97)	1.0
  (8122, 103)	1.0
  (8122, 108)	1.0
  (8123, 5)	1.0
  (8123, 7)	1.0
  (8123, 14)	1.0
  (8123, 21)	1.0
  (8123, 27)	1.0
  (8123, 31)	1.0
  (8123, 33)	1.0
  (8123, 35)	1.0
  (8123, 48)	1.0
  (8123, 49)	1.0
  (8123, 52)	1.0
  (8123, 56)	1.0
  (8123, 63)	1.0
  (8123, 72)	1.0
  (8123, 77)	1.0
  (8123, 79)	1.0
  (8123, 83)	1.0
  (8123, 85)	1.0
  (8123, 93)	1.0
  (8123, 100)	1.0
  (8123, 108)	1.0


In [19]:
# Changing X from sparse representation to standard array
X_train = X.toarray()
X_train

array([[0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 1., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

### Test set

In [42]:
X_test, y_test = get_data("madelon.t")
print(X_test)
print(y_test)

  (0, 0)	483.0
  (0, 1)	454.0
  (0, 2)	513.0
  (0, 3)	495.0
  (0, 4)	523.0
  (0, 5)	469.0
  (0, 6)	453.0
  (0, 7)	477.0
  (0, 8)	506.0
  (0, 9)	479.0
  (0, 10)	470.0
  (0, 11)	510.0
  (0, 12)	463.0
  (0, 13)	423.0
  (0, 14)	511.0
  (0, 15)	500.0
  (0, 16)	494.0
  (0, 17)	493.0
  (0, 18)	472.0
  (0, 19)	477.0
  (0, 20)	497.0
  (0, 21)	503.0
  (0, 22)	474.0
  (0, 23)	472.0
  (0, 24)	483.0
  :	:
  (599, 475)	398.0
  (599, 476)	482.0
  (599, 477)	493.0
  (599, 478)	422.0
  (599, 479)	484.0
  (599, 480)	483.0
  (599, 481)	468.0
  (599, 482)	471.0
  (599, 483)	507.0
  (599, 484)	542.0
  (599, 485)	472.0
  (599, 486)	477.0
  (599, 487)	493.0
  (599, 488)	503.0
  (599, 489)	477.0
  (599, 490)	484.0
  (599, 491)	478.0
  (599, 492)	487.0
  (599, 493)	694.0
  (599, 494)	493.0
  (599, 495)	499.0
  (599, 496)	474.0
  (599, 497)	494.0
  (599, 498)	536.0
  (599, 499)	526.0
[-1. -1. -1.  1. -1.  1. -1. -1. -1.  1.  1.  1. -1.  1.  1. -1. -1.  1.
 -1. -1. -1. -1. -1.  1.  1.  1.  1. -1. -1. -1. -1.  1.

In [86]:
madelon_train = pd.read_csv("madelon_train.data", header=None, sep=" ")
madelon_train = madelon_train.drop(500, axis=1)
X_train = madelon_train.values
len(X_train)

1171

In [85]:
y_train = pd.read_csv("madelon_train.labels", squeeze = True)
y_train = np.array(y_train)
len(y_train)

1999

In [67]:
madelon_valid = pd.read_csv("madelon_valid.data", header=None, sep=" ")
madelon_valid = madelon_valid.drop(500, axis=1)
X_valid = madelon_valid.values
X_valid

array([[483., 454., 513., ..., 498., 523., 510.],
       [485., 508., 493., ..., 495., 548., 477.],
       [483., 521., 507., ..., 465., 508., 503.],
       ...,
       [479., 464., 502., ..., 477., 512., 474.],
       [493., 487., 489., ..., 463., 504., 451.],
       [483., 525., 482., ...,  nan,  nan,  nan]])

In [79]:
y_valid = pd.read_csv("madelon_valid.labels", squeeze=True)
y_valid = np.array(y_valid)
y_valid[5]

-1

## Applying passive agressive online learning

In [25]:
from sklearn.metrics import accuracy_score
from passive_agressive_classic_update import passive_agressive_online

### Classic update

In [78]:
w_classic = passive_agressive_online(X_train, y_train, impl="classic")
w_classic

array([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, na

In [38]:
# Accuracy on the training set
predicted = np.sign(np.dot(X_train, w_classic))
accuracy = accuracy_score(predicted, y_train)
print("%.2f" % (accuracy*100))

94.23


The classic updata performs really well on the training set (94.23%). Let's look at the test set performance